src/data_coordinator/buffered_data_coordinator.cpp

```diff
@@ -141,52 +137,62 @@ void buffered_data_coordinator<TensorDataType>::setup_data_fields(
 
 template <typename TensorDataType>
 int buffered_data_coordinator<TensorDataType>::fetch_to_local_matrix(
-  data_buffer_map_t& buffer_map,
-  const execution_mode mode)
+  const execution_mode mode,
+  data_buffer<IODataType>& buf,
+  El::Int loaded_mini_batch_size,
+  El::Int relative_base_position,
+  int buffer_id)
 {
   generic_data_reader* dr = get_data_reader(mode);
-  int num_parallel_readers = dr->get_num_parallel_readers();
+  dataset& ds = get_dataset(mode);
 
-  prof_region_begin("fetch_to_local_matrix", prof_colors[2], false);
+  std::string prof_title =
+    ("fetch_to_local_matrix " + std::to_string(buffer_id));
+  prof_region_begin(prof_title.c_str(), prof_colors[2], false);
   /// Coordinate all available readers so that they perform I/O in the same step
   /// Check to make sure that the local matrix has space for data
-  data_buffer<IODataType>& buf = get_data_buffer(buffer_map, mode);
 
   buf.m_num_samples_fetched = 0;
-  /// BVE FIXME change the guard
-  if (this->m_comm->get_rank_in_trainer() < num_parallel_readers &&
-      (buf.m_input_buffers[INPUT_DATA_TYPE_SAMPLES]->LocalHeight() != 0 &&
-       buf.m_input_buffers[INPUT_DATA_TYPE_SAMPLES]->LocalWidth() != 0)) {
+  // Check to see if this rank has local space for the current mini-batch
+  if (buf.m_input_buffers[INPUT_DATA_TYPE_SAMPLES]->LocalHeight() != 0 &&
+      buf.m_input_buffers[INPUT_DATA_TYPE_SAMPLES]->LocalWidth() != 0) {
     /// Create a map of the local matrices to pass into the data reader
     std::map<data_field_type, CPUMat*> local_input_buffers;
     for (auto& b : buf.m_input_buffers) {
       local_input_buffers[b.first] =
         static_cast<CPUMat*>(&(b.second->Matrix()));
     }
 
-    // Compute the size of the current mini-batch
-
-    int loaded_batch_size = dr->get_loaded_mini_batch_size();
-    const int end_pos =
-      std::min(static_cast<size_t>(dr->m_current_pos + loaded_batch_size),
-               dr->m_shuffled_indices.size());
-    const int mb_size = std::min(
-      El::Int{((end_pos - dr->m_current_pos) + dr->m_sample_stride - 1) /
-              dr->m_sample_stride},
-      local_input_buffers[INPUT_DATA_TYPE_SAMPLES]->Width());
-
-    /** @brief Each rank will fetch a mini-batch worth of data into it's buffer
+    // Compute the size of the current local mini-batch
+    const int end_pos = std::min(
+      static_cast<size_t>(relative_base_position + loaded_mini_batch_size),
+      dr->m_shuffled_indices.size());
+    const int local_mini_batch_size =
+      std::min(El::Int{((end_pos - relative_base_position) +
+                        ds.get_sample_stride() - 1) /
+                       ds.get_sample_stride()},
+               local_input_buffers[INPUT_DATA_TYPE_SAMPLES]->Width());
```

Comment:
Yeah, I see that a lot of this is inherited code from not-this-PR, but the type management here is a mess.